wip

cupcicm opened the pull request.

cupcicm committed:
Creates a "MetadataRefresh" interface.

And a single flight refresher implementation that fixes this
issue: https://github.com/IBM/sarama/issues/3224

Signed-off-by: Manu C <mcupcic@confluent.io>
Signed-off-by: epot <epot@confluent.io>

aratz-lasa commented on metadata.go:
```diff
@@ -0,0 +1,163 @@
+	r.topics = append(r.topics, topics...)
+}
+
+func (r *metadataRefresh) HasTopics(topics []string) bool {
```
If a refresh for a specific set of topics is already in progress, a new request to refresh all topics will not work correctly. The new request will simply wait for the existing, partial refresh to complete and then return, without ever triggering the "all topics" refresh it asked for

aratz-lasa replied:
We need sth like:

```
// An empty slice means "all topics". If the current refresh is not for all topics, it's a mismatch.
	if len(topics) == 0 {
		return false
	}
```

cupcicm replied:
Updated.

cupcicm committed:
Creates a configuration option to use the new refresher.

It defaults to false, and when you set it to true, sarama
will use a single flight refresher to avoid sending multiple
metadata requests at the same time.

This fixes this issue: https://github.com/IBM/sarama/issues/3224

Signed-off-by: Manu C <mcupcic@confluent.io>
Signed-off-by: epot <epot@confluent.io>

cupcicm committed:
Cosmetic improvements.

1. remove var =
2. use assert in goroutines
3. unexport a bunch of fields
4. some other minor things.

Signed-off-by: Manu C <mcupcic@confluent.io>
Signed-off-by: epot <epot@confluent.io>

cupcicm committed:
Make the tests test the refreshOrQueue method.

Rather than the Refresh method. It's a lower level tests, but it
allows us to write a precise test with no time.Sleep.

Signed-off-by: Manu C <mcupcic@confluent.io>
Signed-off-by: epot <epot@confluent.io>

cupcicm committed:
Address comments

Signed-off-by: Manu C <mcupcic@confluent.io>
Signed-off-by: epot <epot@confluent.io>

epot committed:
Address comments

Signed-off-by: epot <epot@confluent.io>

epot committed:
fixes after review

Signed-off-by: epot <epot@confluent.io>
